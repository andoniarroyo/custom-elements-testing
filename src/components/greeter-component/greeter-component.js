const templateString = `
    <style>
        .greeter-text {
          font-size: 250px;
        }
    </style>
    <div>
      <span class="greeter-text"></span>
      <button id="notice-button">Notice!</button>
    </div>
`;
const template = document.createElement('template');
template.innerHTML = templateString;

export class GreeterComponent extends HTMLElement {
  static nameToGreeteAttributeName = "name-to-greete";
  static tag = "greeter-component";

  static get tag() {
    return tag;
  }

  static get observedAttributes() {
    return [GreeterComponent.nameToGreeteAttributeName];
  }

  #currentName = "...";

  constructor() {
    super();

    this.appendTemplate();
    this.updateGreeterText();
  }

  connectedCallback() {
    this.subscribeToEvents();
  }

  attributeChangedCallback(attrName, oldVal, newVal) {
    this.#currentName = this.getAttribute(GreeterComponent.nameToGreeteAttributeName);
    this.updateGreeterText();
  }

  appendTemplate() {
    let templateContent = template.content;
    document.body.appendChild(templateContent);
  }

  updateGreeterText() {
    document.querySelector(".greeter-text").innerHTML = this.#currentName;
  }

  subscribeToEvents() {
    document.querySelector("#notice-button").addEventListener('click', ()=> {
      this.dispatchEvent(
        new CustomEvent("greetingNoticed", { bubles: true, detail: { noticedName: this.#currentName} }))
    });
  }

}

customElements.define("greeter-component", GreeterComponent);