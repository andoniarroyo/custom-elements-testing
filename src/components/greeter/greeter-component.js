const templateString = `
    <style>
        .greeter-text {
          font-size: 250px;
        }
    </style>
    <div>
      <span class="greeter-text"></span>
      <button id="notice-button">Notice!</button>
    </div>
`;
const template = document.createElement('template');
template.innerHTML = templateString;

export class GreeterComponent extends HTMLElement {
  static nameToGreeteAttributeName = "name-to-greete";
  static tag = "greeter-component";

  static get observedAttributes() {
    return [GreeterComponent.nameToGreeteAttributeName];
  }

  #currentName = "...";

  connectedCallback() {
    this.appendTemplate();
    this.subscribeToEvents();
  }

  appendTemplate() {
    let templateContent = template.content.cloneNode(true);
    this.append(templateContent);
  }

  subscribeToEvents() {
    document.querySelector("#notice-button").addEventListener('click', () => {
      this.dispatchEvent(
        new CustomEvent("greetingNoticed", { bubles: true, detail: { noticedName: this.#currentName} }))
    });
  }

  attributeChangedCallback(attrName, oldVal, newVal) {
    this.#currentName = newVal;
    this.updateGreeterText(newVal);
  }

  updateGreeterText(newGreeterName) {
    document.querySelector(".greeter-text").textContent = newGreeterName;
  }

}

customElements.define("greeter-component", GreeterComponent);