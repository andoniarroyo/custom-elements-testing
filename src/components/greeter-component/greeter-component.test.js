import { Window } from 'happy-dom';
import { GreeterComponent } from "./greeter-component";

const window = new Window();
const document = window.document;

const greeterComponentTag = "greeter-component";
window.customElements.define(greeterComponentTag, GreeterComponent);

describe("The greeter component", () => {
    let element, shadowRoot;
    beforeEach(() => {
        element = document.createElement(greeterComponentTag);
        shadowRoot = element.shadowRoot;
        document.body.appendChild(element);
    });
    
    // A visual component should only deal with:
    // 1 - Create a visual representation of the model
    // 2 - Interact with the user 
    // 3 - Notify to the outside world what relevant actions heppened
    // 
    // I'm interested on check three aspects of the components:
    // 1- DOM generation (it is private)
    // 2- The bindings are properly connected
    // 3- The proper events are raised on the right interactions
    //
    // Everything ending up with a change in the DOM could be covered with snapshot/s
    // To check the events we should create specific tests
    
    describe('on creating', () => {
        it('creates the expected DOM structure', () => {
            // binding of the name to greete (attr -> memory -> DOM)
            const newGreetingValue = "Mr test";
            element.setAttribute("name-to-greete", newGreetingValue);
            expect(document.documentElement.innerHTML).toMatchSnapshot();
        });
    });

    describe('notifying to the outer scope', () => {
        it('on greeting noticed the event is raised including the name of the moment', () => {
            element.addEventListener("greetingNoticed", (payload) => {
                expect(payload).toBeTruthy();
                expect(payload.detail).toBeTruthy();
                expect(payload.detail.noticedName).toBe(newGreetingValue);
            });      

            const newGreetingValue = "Mr test";
            element.setAttribute("name-to-greete", newGreetingValue);
            document.querySelector("#notice-button").click();
        })
    })

})