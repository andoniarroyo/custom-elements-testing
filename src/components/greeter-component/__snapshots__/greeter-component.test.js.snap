// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The greeter component on creating creates the expected DOM structure 1`] = `
"<head></head><body>
    <style>
        .greeter-text {
          font-size: 250px;
        }
    </style>
    <div>
      <span class=\\"greeter-text\\">Mr test</span>
      <button id=\\"notice-button\\">Notice!</button>
    </div>
<greeter-component name-to-greete=\\"Mr test\\"></greeter-component></body>"
`;
